src/devices/sound/ymfm.cpp

```diff
@@ -0,0 +1,1658 @@
+	m_op1.save(device, index * 4 + 0);
+	m_op2.save(device, index * 4 + 1);
+	m_op3.save(device, index * 4 + 2);
+	m_op4.save(device, index * 4 + 3);
+}
+
+
+//-------------------------------------------------
+//  reset - reset the channel state
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_channel<RegisterType>::reset()
+{
+	// reset our data
+	m_feedback[0] = m_feedback[1] = 0;
+
+	// reset the operators
+	m_op1.reset();
+	m_op2.reset();
+	m_op3.reset();
+	m_op4.reset();
+}
+
+
+//-------------------------------------------------
+//  keyonoff - signal key on/off to our operators
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_channel<RegisterType>::keyonoff(u8 states)
+{
+	m_op1.keyonoff(BIT(states, 0));
+	m_op2.keyonoff(BIT(states, 1));
+	m_op3.keyonoff(BIT(states, 2));
+	m_op4.keyonoff(BIT(states, 3));
+}
+
+
+//-------------------------------------------------
+//  keyon_csm - signal CSM key on to our operators
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_channel<RegisterType>::keyon_csm()
+{
+	m_op1.keyon_csm();
+	m_op2.keyon_csm();
+	m_op3.keyon_csm();
+	m_op4.keyon_csm();
+}
+
+
+//-------------------------------------------------
+//  clock - master clock of all operators
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_channel<RegisterType>::clock(u32 env_counter, s8 lfo_raw_pm, bool is_multi_freq)
+{
+	// grab common block/fnum values
+	u16 block_freq = m_regs.block_freq();
+
+	// clock the feedback through
+	m_feedback[0] = m_feedback[1];
+	m_feedback[1] = m_feedback[2];
+
+	// in multi-frequency mode, the first 3 channels use independent block/fnum values
+	if (is_multi_freq)
+	{
+		m_op1.clock(env_counter, lfo_raw_pm, m_regs.multi_block_freq1());
+		m_op2.clock(env_counter, lfo_raw_pm, m_regs.multi_block_freq2());
+		m_op3.clock(env_counter, lfo_raw_pm, m_regs.multi_block_freq0());
+	}
+
+	// otherwise, all channels use the common block/fnum
+	else
+	{
+		m_op1.clock(env_counter, lfo_raw_pm, block_freq);
+		m_op2.clock(env_counter, lfo_raw_pm, block_freq);
+		m_op3.clock(env_counter, lfo_raw_pm, block_freq);
+	}
+
+	// operator 3 uses the common values in all cases
+	m_op4.clock(env_counter, lfo_raw_pm, block_freq);
+}
+
+
+//-------------------------------------------------
+//  output - combine the operators according to the
+//  specified algorithm, returning a sum according
+//  to the rshift and clipmax parameters, which
+//  vary between different OPN implementations
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_channel<RegisterType>::output(u8 lfo_raw_am, u8 noise_state, s32 &lsum, s32 &rsum, u8 rshift, s16 clipmax) const
+{
+	// skip if both pans are clear; no need to do all this work for nothing
+	if (m_regs.pan_left() == 0 && m_regs.pan_right() == 0)
+		return;
```

Comment:
Shouldn’t feedback still be updated if necessary?  It’s not dependent on current pan.

Reply:
You're right about this. I will move the condition to after the feedback is computed.